### ISSUE TITLE
return (NULL) will give incorrect warnings with -Wimplicit-void-ptr-cast

### ISSUE THREAD

**Issue by Leont** [llvm/llvm-project#171874]
This program:
```
#include <stdlib.h>

struct foo {
	char baz;
};

struct foo* bar() {
	return(NULL);
}

int main() {
	bar();
}
```
Will give the following warning when compiled with `-Wimplicit-void-ptr-cast` / `-Wc++-compat`
```
test.c:8:8: warning: implicit conversion when returning 'void *' from a function with result type 'struct foo *' is not
      permitted in C++ [-Wimplicit-void-ptr-cast]
    8 |         return(NULL);
      |               ^~~~~~
1 warning generated.
```
Even though it's perfectly valid C++ (because NULL isn't a void pointer C++, so the expression `(NULL)` won't be a void pointer either)  that clang++ doesn't have any issue with.

**Comment by Leont** [comment]
> The crux is that the warning mechanisms only directly see something like ((void*)0) in C mode, and can't be aware of that NULL expands to different things between C and C++...

It does do the right thing for `return NULL;`, so I imagine there a way to also not make it warn for `return (NULL);`
